File: io/lucamienert/basic/instruction/InstructionSequence.java
package io.lucamienert.basic.instruction;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class InstructionSequence {

    private final List<Instruction> instructions = new ArrayList<>();
    private int labelCounter = 0;

    public String createLineLabel(int line) {
        return "line_" + line;
    }

    public String createGeneratedLabel() {
        return "generated_" + (labelCounter++);
    }

    public void append(Instruction... instructions) {
        this.instructions.addAll(Arrays.asList(instructions));
    }

    public List<Instruction> getInstructions() {
        return Collections.unmodifiableList(instructions);
    }

}

File: BasicCompiler/Instructions/InstructionSequence.cs
namespace BasicCompiler.Instructions;

public class InstructionSequence
{
    private readonly List<Instruction> _instructions = new List<Instruction>();
    private int _labelCounter = 0;

    public string CreateLineLabel(int line) => $"line_{line}";
    public string CreateGeneratedLabel() => $"generated_{_labelCounter++}";
    public void Append(Instruction instruction) => _instructions.Add(instruction);
    public List<Instruction> GetInstructions() => _instructions;
}
